### PR TITLE
feat: delete-single-trace

### LINKED ISSUE #501: feat: delete single trace
### Describe the feature you'd like to request

Delete a single trace

### Describe the solution you'd like to see

- action on tables
- action on single trace view

### Additional information

While usually this is not necessary, there are various reasons why it can be useful to delete a single trace.

### Contribute

- [ ] Yes, I can implement this and raise a PR

### PR DESCRIPTION
**Pull request by TameemAsim**
This PR solves #501 .

After PR:

[Screencast from 11-17-2023 04:40:07 AM.webm](https://github.com/langfuse/langfuse/assets/88766161/199c3fb2-b6fe-4f00-a292-fdaf5bc2ff9f)

### CODE REVIEW

File: src/server/api/routers/traces.ts [hunk 1]
```diff
@@ -275,4 +275,37 @@ export const traceRouter = createTRPCRouter({
         observations: enrichedObservations as ObservationReturnType[],
       };
     }),
+  delete: protectedGetTraceProcedure
```
Comment (marcklingen) [inline src/server/api/routers/traces.ts:278]: This also allows public access if trace is published (meant for reading only). Use protectedProjectProcedure which requires the projectId as a param

File: src/server/api/routers/traces.ts [hunk 2]
```diff
@@ -275,4 +275,37 @@ export const traceRouter = createTRPCRouter({
...
+    .input(z.object({ traceId: z.string() }))
+    .mutation(async ({ input, ctx }) => {
+      const trace = await ctx.prisma.trace.findFirstOrThrow({
```
Comment (marcklingen) [inline src/server/api/routers/traces.ts:281]: add a check whether user has a sufficient role to perform this action

File: src/server/api/routers/traces.ts [hunk 3]
```diff
@@ -275,4 +275,37 @@ export const traceRouter = createTRPCRouter({
...
+        where: {
+          id: input.traceId,
```
Comment (marcklingen) [inline src/server/api/routers/traces.ts:283]: Add check whether the traceId is actually in the project the user has access to

File: src/server/api/routers/traces.ts [hunk 4]
```diff
@@ -275,4 +275,37 @@ export const traceRouter = createTRPCRouter({
...
+        },
+        include: {
+          scores: true
+        }
+      })
+
+      await trace.scores.map(async (score) => {
```
Comment (marcklingen) [inline src/server/api/routers/traces.ts:290]: Scores cascade, need to only delete all observations manually

File: src/components/delete-trace.tsx [hunk 5]
```diff
@@ -0,0 +1,48 @@
...
+  icon?: boolean,
+}
+) {
+  const router = useRouter();
+
+  const mutDeleteTrace = api.traces.delete.useMutation();
+
+  const deleteTrace = async () => {
+    await mutDeleteTrace.mutateAsync({traceId: traceId});
+    if (icon) {
```
Comment (marcklingen) [inline src/components/delete-trace.tsx:23]: if(icon) is unclear, I get why you've implemented it this way but I would rename the icon to eg isTableAction

File: src/components/delete-trace.tsx [hunk 6]
```diff
@@ -0,0 +1,48 @@
...
+      router.reload();
```
Comment (marcklingen) [inline src/components/delete-trace.tsx:24]: Do not reload but use invalidation of trpc cache, thereby the table updates without a full refresh.

File: src/components/delete-trace.tsx [hunk 7]
```diff
@@ -0,0 +1,48 @@
...
+
+  if (icon) {
+    return (
+      <Button variant="ghost" size="xs" onClick={deleteTrace}>
+        <TrashIcon className='w-4 h-4'/>
+      </Button>
+    );
+  };
+
+  return (
```
Comment (marcklingen) [inline src/components/delete-trace.tsx:38]: Add popover to ask for confirmation, this is a dangerous action
UI: https://ui.shadcn.com/docs/components/popover

**Review by marcklingen — CHANGES_REQUESTED** [review]
Thanks a lot for your contribution @18feb06 

I've made some suggestions as this is a dangerous action (need to check for access rights, ask for confirmation by user) and it manipulates the state (invalidate cache instead of refreshing the application)

File: src/server/api/routers/traces.ts (hunk 2 above)
Comment (18feb06) [reply src/server/api/routers/traces.ts:281]: ok... no problem

File: src/server/api/routers/traces.ts (hunk 4 above)
Comment (18feb06) [reply src/server/api/routers/traces.ts:290]: ok... no problem

File: src/components/delete-trace.tsx (hunk 5 above)
Comment (18feb06) [reply src/components/delete-trace.tsx:23]: ok... no problem

File: src/components/delete-trace.tsx (hunk 7 above)
Comment (18feb06) [reply src/components/delete-trace.tsx:38]: ok... no problem  working on all issues

File: src/components/delete-trace.tsx (hunk 6 above)
Comment (18feb06) [reply src/components/delete-trace.tsx:24]: ok... no problem

File: src/server/api/routers/traces.ts (hunk 2 above)
Comment (18feb06) [reply src/server/api/routers/traces.ts:281]: I have checked scopes array, there is no scope for trace:CUD or trace:delete.
I am thinking to create new scopes. Kindly give any suggestion if you have.
If yes, then tell me that should only admin and owner be allowed to delete traces or member can also delete traces?

File: src/server/api/routers/traces.ts (hunk 3 above)
Comment (marcklingen) [reply src/server/api/routers/traces.ts:290]: Moved this to the sql query

Comment (18feb06) [reply src/server/api/routers/traces.ts:290]: ok

**Review by marcklingen — APPROVED** [review]
Thanks a lot for the contribution, I made minor tweaks

- delete observations and trace in one transaction
- hide delete button for users who do not have access
- change text on popup
- change invalidation/redirect logic to not cause any erroring trpc calls

**Pull request merged**